eclipse-theia/theia
# Enable Workspace-Scoped Tasks

#### What it does

Fixes #8519
Fixes #8991 (thanks for noting that, @RomanNikitenko)
This PR sets up the necessary infrastructure to read and write tasks to a workspace file. In particular it:
 - modifies the way a `WorkspaceFilePreferenceProvider` reads and writes non-settings preferences. (breaking changes to bring Theia's workspace file schema into harmony with VSCode's)
 - modifies helper methods in `workspace-service.ts` for writing workspace files to accommodate fields other than `folders` and `settings` (breaking changes for anyone calling those file-builder methods)
 - removes some explicit references to the preference system in the workspace package.
 - adds a `WorkspaceSchemaUpdater` service to allow packages that contribute extensions to the schema of the workspace file (tasks, debug) to register their extensions
 - injects workspace scope related functionality into the `TaskConfigurationManager`

#### How to test

1. Add a properly formatted `tasks` field to a workspace file. (`"tasks": {"tasks": [...]}` - has be double `"tasks"`, one for the 'file' label (`tasks.json`) and one for the tasks inside)
2. Open that file as your workspace.
3. Open the `Terminal` menu and select `Run Task...`
4. Observe that your tasks are available and run (or fail) as expected
5. Modify the file (e.g. add a new task configuration)
6. Repeat 3-4, and observe that the task list updated correctly.
---
1. Use task quick picks (commands `tasks configure`, `tasks run`) and confirm that they still work in each scope in single- and multi-root workspaces.
> NB: running very short-lived tasks (e.g. a `shell` task that just runs `echo`) can behave unexpectedly, with no terminal ever opening. This behavior is also found on `master`.

#### Changes

This PR has grown to touch quite a number of files, so it might be useful for reviewers if I explain why each change was necessary:
- Core
  - `preference-configurations.ts`: add an `isAnyConfig` convenience method equivalent to checking `isSectionName() || isConfigName()`
  - `preference-contribution.ts`: change schema update logic to respect the superset / subset relationships among scopes. See https://github.com/eclipse-theia/theia/pull/8917#issuecomment-767592610.
  - `preference-provider.ts`, `preference-service.ts`: add section name argument to `getConfigUri` and `getContainingConfigUri` to allow callers to find the URI of all preference files, not just `settings.json`. This is more systematic than the current implementation, and obviates workarounds using `PreferenceService.resolve` to find the URI for non-settings sections.
- Debug
  - `debug-schema-updater.ts`: add `launch` to the workspace file schema.
- Preferences
  - `folders-preference-provider.ts`, `user-configs-preference-provider.ts`, `workspace-preference-provider`: update `getConfigUri` and `getContainingConfigUri` to accommodate section name.
  - `section-preference-provider.ts`: minor cleanup to remove a non-null assertion.
  - `workspace-file-preference-provider.ts`: changes to handle sections other than `settings`, and to accommodate sections both inside the `settings` object and outside.
- Tasks
  - `quick-open-tasks.ts`: refactor the `configure` method to accommodate workspace-scoped tasks.
  - `task-configuration-manager.ts`: rewrite logic to accommodate workspace-scoped tasks.
  - `task-configuration-model.ts`: make scope protected because in the case of workspace scope it becomes unreliable (sometimes pointing to folder, sometimes pointing to workspace file). It wasn't being accessed, and the only way to get access to a model is to know the scope you're interested in, so you shouldn't have to ask the model for that information.
  - `task-configurations.ts`: change to allow configure to work in workspace scope.
  - `task-preferences.ts`: better default value for preferences.
  - `task-schema-updater.ts`: make the basic schema for tasks publicly accessible (it is for preferences) so that it can be used for any defaulting. Also add tasks the workspace-file schema.
- Workspace
  - `workspace-commands.ts`, `workspace-frontend-contribution.ts`: add a command to open the workspace file. It exists in VSCode and is parallel to commands to open e.g. the user-scope `tasks.json`.
  - `workspace-schema-updater.ts`, `workspace-frontend-module.ts`, `workspace-service.ts`: add a `WorkspaceSchemaUpdater` that can be accessed and updated by external packages (debug, tasks, preferences) to allow them to register their subschemas for the workspace file. Modify helper functions that assume that the schema will only ever include `settings`.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
